File: LinkedList/python/problem_160_intersection_of_two_linked_lists.py
"""Problem 160: Intersection of Two Linked Lists - Easy
Find the node where two linked lists intersect
Approach: Two pointers - redirect to other list when reaching end
Time: O(m + n), Space: O(1)
"""
class ListNode:
    def __init__(self, val=0, next=None):
        self.val, self.next = val, next

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if not headA or not headB:
            return None
        
        pA, pB = headA, headB
        
        # When pointers meet, they're either at intersection or both None
        while pA != pB:
            pA = headB if pA is None else pA.next
            pB = headA if pB is None else pB.next
        
        return pA

File: LinkedList/java/Problem160_IntersectionOfTwoLinkedLists.java
/**
 * Problem 160: Intersection of Two Linked Lists
 * Difficulty: Easy
 * 
 * Find the node where two linked lists intersect.
 * 
 * Approach: Two pointers - when one reaches end, redirect to other list's head
 * Time: O(m + n), Space: O(1)
 */
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }
}

public class Problem160_IntersectionOfTwoLinkedLists {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;
        
        ListNode pA = headA;
        ListNode pB = headB;
        
        // When pointers meet, they're either at intersection or both null
        while (pA != pB) {
            pA = (pA == null) ? headB : pA.next;
            pB = (pB == null) ? headA : pB.next;
        }
        
        return pA;
    }
}
